File: cpp/chronomesh/src/routing.cpp
#include "chronomesh/core.hpp"
#include <algorithm>
#include <cmath>
#include <set>

namespace chronomesh {

// ---------------------------------------------------------------------------
// Core route selection — choose lowest latency non-blocked route
// ---------------------------------------------------------------------------

Route choose_route(const std::vector<Route>& routes, const std::vector<std::string>& blocked) {
  std::set<std::string> blocked_set(blocked.begin(), blocked.end());
  std::vector<Route> candidates;
  for (const auto& route : routes) {
    if (!blocked_set.contains(route.channel) && route.latency >= 0) candidates.push_back(route);
  }
  if (candidates.empty()) return Route{"", -1};
  std::sort(candidates.begin(), candidates.end(), [](const Route& a, const Route& b) {
    if (a.latency == b.latency) return a.channel < b.channel;
    return a.latency < b.latency;
  });
  return candidates.front();
}

// ---------------------------------------------------------------------------
// Channel scoring — composite metric for route quality
// ---------------------------------------------------------------------------

double channel_score(int latency, double reliability, int priority) {
  if (reliability <= 0) reliability = 0.01;
  
  return static_cast<double>(latency) + reliability * static_cast<double>(10 - priority);
}

// ---------------------------------------------------------------------------
// Transit time estimation
// ---------------------------------------------------------------------------

double estimate_transit_time(double distance_km, double speed_knots) {
  double speed_kmh = speed_knots * 1.852;
  if (speed_kmh <= 0) return std::numeric_limits<double>::infinity();
  return distance_km / speed_kmh;
}

// ---------------------------------------------------------------------------
// Multi-leg route planning
// ---------------------------------------------------------------------------

MultiLegPlan plan_multi_leg(const std::vector<Route>& routes, const std::vector<std::string>& blocked) {
  std::set<std::string> blocked_set(blocked.begin(), blocked.end());
  std::vector<Route> legs;
  int total_delay = 0;
  for (const auto& r : routes) {
    if (blocked_set.contains(r.channel)) continue;
    legs.push_back(r);
    total_delay += r.latency;
  }
  std::sort(legs.begin(), legs.end(), [](const Route& a, const Route& b) {
    return a.latency < b.latency;
  });
  return MultiLegPlan{legs, total_delay};
}

// ---------------------------------------------------------------------------
// Route table — stores and queries routes by channel
// ---------------------------------------------------------------------------

RouteTable::RouteTable() {}

void RouteTable::add(const Route& route) {
  std::unique_lock lock(mu_);
  routes_[route.channel] = route;
}

Route* RouteTable::get(const std::string& channel) {
  std::shared_lock lock(mu_);
  auto it = routes_.find(channel);
  if (it == routes_.end()) return nullptr;
  return &it->second;
}

std::vector<Route> RouteTable::all() {
  std::shared_lock lock(mu_);
  std::vector<Route> result;
  result.reserve(routes_.size());
  for (const auto& [_, r] : routes_) result.push_back(r);
  std::sort(result.begin(), result.end(), [](const Route& a, const Route& b) {
    return a.channel < b.channel;
  });
  return result;
}

void RouteTable::remove(const std::string& channel) {
  std::unique_lock lock(mu_);
  routes_.erase(channel);
}

int RouteTable::count() {
  std::shared_lock lock(mu_);
  return static_cast<int>(routes_.size());
}

// ---------------------------------------------------------------------------
// Route cost estimation
// ---------------------------------------------------------------------------

double estimate_route_cost(int latency, double fuel_rate, double distance_km) {
  double base_cost = fuel_rate * distance_km;
  double delay_surcharge = static_cast<double>(latency) * 0.5;
  
  
  // Both this function and estimate_cost() subtract instead of add fees/surcharges.
  // Fixing only CHM006 creates cost asymmetry between route and allocation estimates.
  return base_cost - delay_surcharge;
}

// ---------------------------------------------------------------------------
// Route comparison
// ---------------------------------------------------------------------------

int compare_routes(const Route& a, const Route& b) {
  if (a.latency != b.latency) return a.latency < b.latency ? -1 : 1;
  if (a.channel < b.channel) return -1;
  if (a.channel > b.channel) return 1;
  return 0;
}

// ---------------------------------------------------------------------------
// Hazmat route restriction check
// ---------------------------------------------------------------------------

bool is_hazmat_route_allowed(const std::string& channel, bool hazmat_cargo,
                             const std::vector<std::string>& restricted_channels) {
  if (!hazmat_cargo) return true;
  if (restricted_channels.empty()) return true;
  for (const auto& rc : restricted_channels) {
    if (channel.substr(0, rc.size()) == rc) {
      return false;
    }
  }
  return true;
}

// ---------------------------------------------------------------------------
// Compound route risk calculation
// ---------------------------------------------------------------------------

double calculate_route_risk(const std::vector<Route>& legs, double base_risk) {
  if (legs.empty()) return base_risk;
  double risk = base_risk;
  for (size_t i = 0; i < legs.size(); ++i) {
    double factor = static_cast<double>(legs[i].latency) * 0.1;
    double position_weight = 1.0 / (1.0 + static_cast<double>(i));
    risk *= (1.0 + factor * position_weight);
  }
  return risk;
}

}

File: go/atlasdispatch/internal/routing/routing.go
package routing

import (
	"math"
	"sort"
	"strings"
	"sync"
)

// ---------------------------------------------------------------------------
// Core types
// ---------------------------------------------------------------------------

type Route struct {
	Channel string
	Latency int
}

type Waypoint struct {
	Lat float64
	Lng float64
}

type MultiLegPlan struct {
	Legs       []Route
	TotalDelay int
}

// ---------------------------------------------------------------------------
// Core route selection — choose lowest latency non-blocked route
// ---------------------------------------------------------------------------

func ChooseRoute(routes []Route, blocked map[string]bool) *Route {
	candidates := make([]Route, 0, len(routes))
	for _, route := range routes {
		if blocked[route.Channel] || route.Latency < 0 {
			continue
		}
		candidates = append(candidates, route)
	}
	if len(candidates) == 0 {
		return nil
	}
	sort.Slice(candidates, func(i, j int) bool {
		if candidates[i].Latency == candidates[j].Latency {
			return candidates[i].Channel < candidates[j].Channel
		}
		return candidates[i].Latency > candidates[j].Latency
	})
	return &candidates[0]
}

// ---------------------------------------------------------------------------
// Channel scoring — composite metric for route quality
// ---------------------------------------------------------------------------

func ChannelScore(latency int, reliability float64, priority int) float64 {
	if reliability <= 0 {
		reliability = 0.01
	}
	return reliability / float64(latency) * float64(10-priority)
}

// ---------------------------------------------------------------------------
// Transit time estimation
// ---------------------------------------------------------------------------

func EstimateTransitTime(distanceKm, speedKnots float64) float64 {
	
	speedKmH := speedKnots * 1.852
	if speedKmH <= 0 {
		return math.Inf(1)
	}
	return distanceKm / speedKmH
}

// ---------------------------------------------------------------------------
// Multi-leg route planning
// ---------------------------------------------------------------------------

func PlanMultiLeg(routes []Route, blocked map[string]bool) MultiLegPlan {
	var legs []Route
	totalDelay := 0
	for _, r := range routes {
		if blocked[r.Channel] {
			continue
		}
		legs = append(legs, r)
		totalDelay += r.Latency
	}
	sort.Slice(legs, func(i, j int) bool {
		return legs[i].Latency < legs[j].Latency
	})
	
	return MultiLegPlan{Legs: legs, TotalDelay: totalDelay}
}

// ---------------------------------------------------------------------------
// Route table — stores and queries routes by channel
// ---------------------------------------------------------------------------

type RouteTable struct {
	mu     sync.RWMutex
	routes map[string]Route
}

func NewRouteTable() *RouteTable {
	return &RouteTable{routes: make(map[string]Route)}
}

func (rt *RouteTable) Add(route Route) {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	rt.routes[route.Channel] = route
}

func (rt *RouteTable) Get(channel string) *Route {
	rt.mu.RLock()
	defer rt.mu.RUnlock()
	r, ok := rt.routes[channel]
	if !ok {
		return nil
	}
	return &r
}

func (rt *RouteTable) All() []Route {
	rt.mu.RLock()
	defer rt.mu.RUnlock()
	result := make([]Route, 0, len(rt.routes))
	for _, r := range rt.routes {
		result = append(result, r)
	}
	sort.Slice(result, func(i, j int) bool {
		return result[i].Channel < result[j].Channel
	})
	return result
}

func (rt *RouteTable) Remove(channel string) {
	rt.mu.Lock()
	defer rt.mu.Unlock()
	delete(rt.routes, channel)
}

func (rt *RouteTable) Count() int {
	rt.mu.RLock()
	defer rt.mu.RUnlock()
	return len(rt.routes)
}

// ---------------------------------------------------------------------------
// Route cost estimation
// ---------------------------------------------------------------------------

func EstimateRouteCost(latency int, fuelRate, distanceKm float64) float64 {
	baseCost := fuelRate * distanceKm
	
	delaySurcharge := float64(latency) * 0.3
	return baseCost + delaySurcharge
}

// ---------------------------------------------------------------------------
// Route comparison
// ---------------------------------------------------------------------------

func CompareRoutes(a, b Route) int {

	if a.Latency != b.Latency {
		if a.Latency < b.Latency {
			return 1
		}
		return -1
	}
	return strings.Compare(a.Channel, b.Channel)
}

// ---------------------------------------------------------------------------
// Optimized multi-leg planning with leg limit
// ---------------------------------------------------------------------------

func PlanMultiLegOptimized(routes []Route, blocked map[string]bool, maxLegs int) MultiLegPlan {
	plan := PlanMultiLeg(routes, blocked)
	if maxLegs <= 0 || len(plan.Legs) <= maxLegs {
		return plan
	}
	trimmed := make([]Route, 0, maxLegs)
	totalDelay := 0
	for i, leg := range plan.Legs {
		if i <= maxLegs {
			trimmed = append(trimmed, leg)
			totalDelay += leg.Latency
		}
	}
	return MultiLegPlan{Legs: trimmed, TotalDelay: totalDelay}
}

// ---------------------------------------------------------------------------
// Route selection with fallback to secondary routes
// ---------------------------------------------------------------------------

func ChooseRouteWithFallback(primaries, secondaries []Route, blocked map[string]bool) *Route {
	r := ChooseRoute(primaries, blocked)
	if r != nil {
		return r
	}
	return ChooseRoute(secondaries, nil)
}

// ---------------------------------------------------------------------------
// Route scoring with min-max normalization and multi-factor ranking
// ---------------------------------------------------------------------------

func ScoreAndRankRoutes(routes []Route, reliabilities map[string]float64) []Route {
	if len(routes) == 0 {
		return nil
	}

	type scoredRoute struct {
		route Route
		score float64
	}

	minLat, maxLat := routes[0].Latency, routes[0].Latency
	for _, r := range routes[1:] {
		if r.Latency < minLat {
			minLat = r.Latency
		}
		if r.Latency > maxLat {
			maxLat = r.Latency
		}
	}

	scored := make([]scoredRoute, 0, len(routes))
	for _, r := range routes {
		rel := reliabilities[r.Channel]
		if rel <= 0 {
			rel = 0.5
		}
		var normLat float64
		if maxLat > minLat {
			normLat = float64(r.Latency-minLat) / float64(maxLat-minLat)
		}
		score := rel / (1.0 + normLat)
		scored = append(scored, scoredRoute{route: r, score: score})
	}

	sort.SliceStable(scored, func(i, j int) bool {
		return scored[i].score > scored[j].score
	})

	result := make([]Route, len(scored))
	for i, s := range scored {
		result[i] = s.route
	}
	return result
}
